external-crates/move/crates/move-bytecode-utils/src/lib.rs

```diff
@@ -47,9 +46,39 @@ impl<'a> Modules<'a> {
         self.iter_modules().into_iter().cloned().collect()
     }
 
-    /// Compute a dependency graph for `self`
-    pub fn compute_dependency_graph(&self) -> DependencyGraph {
-        DependencyGraph::new(self.0.values().copied())
+    /// Return an iterator over the modules in `self` in topological order--modules with least deps first.
+    /// Fails with an error if `self` contains circular dependencies.
+    /// Tolerates missing dependencies.
+    pub fn compute_topological_order(&self) -> Result<impl Iterator<Item = &CompiledModule>> {
+        let mut module_id_idx_map = HashMap::new();
+        let mut idx_module_map = HashMap::new();
+        for (i, m) in self.iter_modules().into_iter().enumerate() {
+            if module_id_idx_map.insert(m.self_id(), i).is_some() {
+                panic!("Duplicate module found")
+            };
+            idx_module_map.insert(i, m);
+        }
+
+        let mut graph: DiGraphMap<usize, usize> = DiGraphMap::new();
+        for i in 0..idx_module_map.len() {
+            graph.add_node(i);
+        }
+
+        for (i, m) in idx_module_map.iter() {
+            for dep in m.immediate_dependencies() {
+                if let Some(j) = module_id_idx_map.get(&dep) {
+                    graph.add_edge(*i, *j, 0);
+                }
+            }
+        }
+
+        match petgraph::algo::toposort(&graph, None) {
+            Err(_) => panic!("Circular dependency detected"),
```

Comment:
It was like that in the old code too. My impression is that you should never be able to end up with a circular dependency on what is essentially a `CompiledModule` list. Because otherwise these modules wouldn't even compile. So it's not like a user error here but some serious bug somewhere else so a panic seems fine here.